integration/upload_test.go

```diff
@@ -0,0 +1,250 @@
+	CorrelationID string            `json:"correlationId"`
+	Options       map[string]string `json:"options"`
+}
+
+type uploadStatus struct {
+	State    string `json:"state"`
+	Progress int    `json:"progress"`
+}
+
+func (suite *uploadTestSuite) SetupSuite() {
+	suite.initializer = util.SuiteInitializer{}
+	suite.initializer.Setup(suite.T())
+
+	opts := env.Options{RequiredIfNoDef: false}
+	suite.cfg = uploadTestConfig{}
+	require.NoError(suite.T(), env.Parse(&suite.cfg, opts), "failed to process environment variables")
+	suite.checkUploadDir()
+
+	thingCfg, err := util.GetThingConfiguration(suite.initializer.Cfg, suite.initializer.MQTTClient)
+	require.NoError(suite.T(), err, "failed to get thing configuration")
+
+	suite.thingURL = fmt.Sprintf("%s/api/2/things/%s", strings.TrimSuffix(suite.initializer.Cfg.DigitalTwinAPIAddress, "/"),
+		thingCfg.DeviceID)
+	suite.featureURL = fmt.Sprintf("%s/features/%s", suite.thingURL, featureID)
+}
+
+func (suite *uploadTestSuite) TearDownSuite() {
+	suite.initializer.TearDown()
+}
+
+func TestFileUpload(t *testing.T) {
+	suite.Run(t, new(uploadTestSuite))
+}
+
+func (suite *uploadTestSuite) TestHTTPUpload() {
+	if len(suite.cfg.HTTPServer) == 0 {
+		suite.T().Skip("HTTP_SERVER variable must be set")
+	}
+	upload := newHTTPUpload(suite.T(), suite.cfg.HTTPServer)
+	suite.testUpload(upload)
+}
+
+func (suite *uploadTestSuite) TestAzureUpload() {
+	options, err := uploaders.GetAzureTestOptions(suite.T())
+	require.NoError(suite.T(), err, "error getting azure test options")
+	upload := newAzureUpload(suite.T(), options)
+	suite.testUpload(upload)
+}
+
+func (suite *uploadTestSuite) TestAWSUpload() {
+	options := uploaders.GetAWSTestOptions(suite.T())
+	upload, err := newAWSUpload(suite.T(), options)
+	require.NoError(suite.T(), err, "error creating AWS client")
+	suite.testUpload(upload)
+}
+
+func (suite *uploadTestSuite) triggerUploads() map[string]string {
+	requestPath := fmt.Sprintf("/features/%s/outbox/messages/request", featureID)
+	requestedFiles := make(map[string]string)
+
+	connMessages, err := util.NewDigitalTwinWSConnection(suite.initializer.Cfg)
+	require.NoError(suite.T(), err, "Failed to create websocket connection")
+	defer connMessages.Close()
+
+	suite.startListening(connMessages, typeMessages)
+	url := fmt.Sprintf("%s/inbox/messages/%s", suite.featureURL, operationTrigger)
+	_, err = util.SendDigitalTwinRequest(suite.initializer.Cfg, http.MethodPost, url, map[string]interface{}{
+		paramCorrelationID: "test",
+	})
+	require.NoError(suite.T(), err, "error sending digital twin request for trigger operation")
+	err = util.ProcessWSMessages(suite.initializer.Cfg, connMessages,
+		func(event *protocol.Envelope) (bool, error) {
+			if requestPath == event.Path {
+				request := &uploadRequest{}
+				err := parseEventValue(event.Value, request)
+				if err != nil {
+					suite.T().Logf("cannot convert %v to upload request, will ignore the event", event.Value)
+					return false, nil
```

Comment:
General topic, the suite connector doesn't use the callback in this way. It has to be aligned.